---
title: "Pan-cancer dual TCR analysis"
author: "Laura Rogers, PhD"
contact: "Rogers.Laura@mayo.edu"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: TRUE
    toc_float:
        collapsed: FALSE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, results = "hide")
```

## Purpose of analysis
In 2021, Zheng et. al published a [pan-cancer single cell atlas](https://www.science.org/doi/10.1126/science.abe6474?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed#supplementary-materials) with gene expression data from 316 cancer patients and TCRseq data from 87 cancer patients. Tissues included in studies include peripheral blood, normal tissue, and tumor tissue. Authors used this data for repertoire analysis comparisons and to identify the routes to T cell exhaustion. However, authors did not evaluate the frequency of cells expressing dual productive TCR sequences. Here, we re-analyze the Zheng Atlas TCRseq data to compute the frequencies of dual TCR expressing T cells in cancer and identify differences in clonal expansion between tissue type and T cell phenotype.


### Package requirements
```{r Package Requirements, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggplot2)
library(jmv)
```




### Data import
The Zhang lab provides objects containing pre-processed [data](https://zenodo.org/record/5461803#.Y5D4sOzMKAk). These were downloaded and read into R. Gene expression data were integrated by the original authors using the [Harmony](https://www.nature.com/articles/s41592-019-0619-0) correction algorithm.  

* data/tcr/byCell/tcr.zhangLab.comb.flt.rds  
  * A data.table containing TCR sequences generated by 10X V(D)J and Smart-Seq2 approaches. One line per cell.  

```{r Data Import, message=FALSE, warning=FALSE}
tcr <- readRDS("data/tcr/byCell/tcr.zhangLab.comb.flt.rds")
```




### TCR data preprocessing
Preprocessing: Some cells (unique identifier = "Cell_Name") have a value in both CDR3.A1 and CDR3.A2 (and/or B1 and B2). Additionally, some fields have NA values, or a string "Couldn't find FGXG". The following code subsets data into 4 dataframes by number of TCR alleles: dualA_tcr, dualB_tcr, dualAB_tcr, single_tcr.

```{r TCR Data Preprocessing, collapse=TRUE}
# This subsets the tcr dataframe for any cell that has 2 TCRa amino acid sequences. (Includes cells with any number of TCRb sequences)
dualA_prep <- subset(tcr, subset = (!is.na(tcr$CDR3.A1) & 
                               !is.na(tcr$CDR3.A2) &
                               !grepl("find|-", tcr$CDR3.A1) &
                               !grepl("find|-", tcr$CDR3.A2)))

# This subsets the tcr dataframe for any cell that has 2 TCRb amino acid sequences. (Includes cells with any number of TCRa sequences)
dualB_prep <- subset(tcr, subset = (!is.na(tcr$CDR3.B1) & 
                               !is.na(tcr$CDR3.B2) &
                               !grepl("find|-", tcr$CDR3.B1) &
                               !grepl("find|-", tcr$CDR3.B2)))

# This subsets the tcr dataframe for any cell that has 2 TCRa amino acid sequences AND 2 TCRb amino acid sequences.
dualAB <- inner_join(dualA_prep, dualB_prep, by = "Cell_Name") 
dualAB_list <- dualAB$Cell_Name #get cell names of the 2927 cells with 2 TCRa and 2 TCRb
dualAB_tcr <- subset(tcr, subset = tcr$Cell_Name %in% dualAB_list) #cells with 2 TCRa and 2 TCRb

# Prepare dual TCR dataframes 
dualA_tcr <- anti_join(dualA_prep, dualAB_tcr, by = "Cell_Name") #cells with 2 TCRa and 1 TCRb
dualB_tcr <- anti_join(dualB_prep, dualAB_tcr, by = "Cell_Name") #cells with 1 TCRa and 2 TCRb

# Prepare single TCR dataframe
anyDualCell_list <- c(dualA_tcr$Cell_Name, dualB_tcr$Cell_Name, dualAB_tcr$Cell_Name)
single_tcr <- subset(tcr, subset = !(tcr$Cell_Name %in% anyDualCell_list))

dim(dualA_tcr) #15,754 cells have 2 TCRa and 1 TCRb
dim(dualB_tcr) #5,381 cells have 1 TCRa and 2 TCRb
dim(dualAB_tcr) #2,927 cells have 2 TCRa and 2 TCRb. Possible doublets?
dim(single_tcr) #144,839,974 cells have 1 TCRa and 1 TCRb

sum(dim(dualA_tcr)[1], dim(dualB_tcr)[1], dim(dualAB_tcr)[1], dim(single_tcr)[1]) == dim(tcr)[1] #Sanity check that all cells are included in dataframes
```



### Dual TCR annotation
This code makes a new master dataframe that includes a new column with TCR allele count annotations:  "Dual TCRa", "Dual TCRa and TCRb", "Dual TCRb" and "Single TCR".

```{r Dual TCR annotation, collapse=TRUE}
# create new master tcr dataframe with Allele_Count annotation column
dualA_tcr$allele_count <- "Dual TCRa"
dualB_tcr$allele_count <- "Dual TCRb"
dualAB_tcr$allele_count <- "Dual TCRa and TCRb"
single_tcr$allele_count <- "Single TCR"
tcr_annot <- bind_rows(dualA_tcr, dualB_tcr, dualAB_tcr, single_tcr)
tcr_annot_noDoublet <- bind_rows(dualA_tcr, dualB_tcr, single_tcr)
```




## Results

### Frequency of dual TCR
There are a total of 168,901 cells with TCRs across all samples. 85.8% of these are expressing a single alpha and beta TCR allele that is expected to be productive (translated). Of the remaining **14.2%** of cells expressing more than one alpha or beta chain allele, the majority of these expressed two alpha chain alleles and one beta chain allele (9.3%), while 3.2% expressed one alpha chain allele and two beta chain alleles. This is consistent with prior observations where co-expression of two alpha chain alleles was more frequently observed than beta chain alleles. Cells expressing two alpha chain and two beta chain alleles (1.7% of cells in dataset) were excluded because it was not possible to determine whether these were doublets that made it through initial quality control settings or true expression values.

```{r Frequency all T cells, echo=FALSE, fig.show='hold', out.width="50%"}
# Count frequency of dual TCR cells
numCells <- table(tcr_annot_noDoublet$allele_count)

# Visualize as cell count bargraph
barData <- data.frame(Allele_Count = names(numCells),
                      Num_Cells = unlist(unname(numCells)))

ggplot(barData, aes(x=Allele_Count, y=Num_Cells.Freq, fill=Allele_Count)) +
  geom_bar(stat = "identity", width = 0.6, show.legend = FALSE) +
  theme_minimal() +
  scale_fill_manual(values = c("paleturquoise4", "lightslateblue", "gray87")) +
  ylab("Number of Cells") +
  xlab(NULL) +
  ggtitle("Total T cell counts") +
  theme(plot.title = element_text(hjust = 0.5, size = 14),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank()) 
ggsave("Fig.1A Total T cell counts.pdf", device = "pdf", path = "output", width = 4, height = 3)

# Calculate proportion of dual TCR cells in dataset
barData <- data.frame(Allele_Count = names(numCells),
                      Num_Cells = 100*(unlist(unname(numCells))/sum(unlist(unname(numCells)))))
```




### Frequency of unique clones
There are 95,102 unique TCR clones from 168,901 total T cells in this dataset. 82.6% of unique clones are expressing a single alpha and beta TCR allele that is expected to be productive (translated). Of the remaining **17.4%** of unique clones expressing more than one alpha or beta chain allele, the majority of these (10.6% of unique clones) are expressing two alpha chain alleles and one beta chain allele.

```{r unique_tcr master, echo=FALSE}

# select rows with first instance of unique cloneID 
dA_uniqClones <- subset(dualA_tcr, subset = !duplicated(dualA_tcr$cloneID)) #This subsets unique cloneIDs
dB_uniqClones <- subset(dualB_tcr, subset = !duplicated(dualB_tcr$cloneID))
dAB_uniqClones <- subset(dualAB_tcr, subset = !duplicated(dualAB_tcr$cloneID))
single_uniqClones <- subset(single_tcr, subset = !duplicated(single_tcr$cloneID))

# create new unique tcr dataframe with Allele_Count annotation column
dA_uniqClones$allele_count <- "Dual TCRa"
dB_uniqClones$allele_count <- "Dual TCRb"
dAB_uniqClones$allele_count <- "Dual TCRa and TCRb"
single_uniqClones$allele_count <- "Single TCR"
unique_tcr_annot <- bind_rows(dA_uniqClones, dB_uniqClones, dAB_uniqClones, single_uniqClones)
unique_tcr_annot_noDoublet <- bind_rows(dA_uniqClones, dB_uniqClones, single_uniqClones)
```

```{r freq unique clones, echo=FALSE, fig.show='hold', out.width="50%"}

# Count dual TCR clones
numClones <- table(unique_tcr_annot_noDoublet$allele_count)

# Visualize as clone count bargraph
barData <- data.frame(Allele_Count = names(numClones),
                      Num_Clones = unlist(unname(numClones)))

ggplot(barData, aes(x=Allele_Count, y=Num_Clones.Freq, fill=Allele_Count)) +
  geom_bar(stat = "identity", width = 0.6, show.legend = FALSE) +
  theme_minimal() +
  scale_fill_manual(values = c("paleturquoise4", "lightslateblue", "gray87")) +
  ylab("Number of Clones") +
  xlab(NULL) +
  ggtitle("T cell unique clone counts") +
  theme(plot.title = element_text(hjust = 0.5, size = 14),
        axis.ticks.x = element_blank(),
        panel.grid.minor = element_blank()) 
ggsave("Fig.1B T cell unique clone counts.pdf", device = "pdf", path = "output", width = 4, height = 3)
```




### Clonal expansion
Using the pre-calculated clone size measures, clone sizes were compared across TCR allele count groups.

```{r prepDensity function, echo=FALSE}
prepDensity <- function(dfA, dfB, dfS) {
  # This function takes 3 dataframes as input and prepares the cloneSize column data 
  # in long format for graphing and statistical analysis
  
  a <- dfA$cloneSize
  b <- dfB$cloneSize
  s <- dfS$cloneSize
  
  maxLength <- max(length(a), length(b), length(s))
  length(a) <- maxLength #Fill list to longest list length with NAs
  length(b) <- maxLength
  length(s) <- maxLength
  
  df <- data.frame("dual TCRa" = a,
                   "dual TCRb" = b,
                   "single TCR" = s)
  df_long <- na.omit(df %>% gather(key = "Allele_Count", value = "Clone_Size"))
  return(df_long)
}
```

#### All unique clones
The mean clone size of all three dual TCR categories is significantly greater (range 1.6X - 4.8X) than the clone size of single TCR expressing clones. This suggests that a greater number of individual clones in dual expressing categories are expanded.

```{r prop unique clonally expanded, echo=FALSE, warning=FALSE, fig.show='hold', out.width="50%"}

# What is the proportion of unique clones within allele groups that are clonally expanded?
da_propClonal <- table(dA_uniqClones$cloneStatus)[1] / dim(dA_uniqClones)[1]
db_propClonal <- table(dB_uniqClones$cloneStatus)[1] / dim(dB_uniqClones)[1]
s_propClonal <- table(single_uniqClones$cloneStatus)[1] / dim(single_uniqClones)[1]
  

# What is the mean clone size in each category?
## Generate summary statistics for unique clone sizes
summary_CloneSize <- lapply(list("dual TCRa" = dA_uniqClones$cloneSize, 
                                 "dual TCRb" = dB_uniqClones$cloneSize, 
                                 "single TCR" = single_uniqClones$cloneSize),
                            summary)
summary_CloneSize


cloneSizes_allClones <- prepDensity(dfA = dA_uniqClones, dfB = dB_uniqClones, dfS = single_uniqClones)

## Choosing statistical test
#bartlett.test(Clone_Size ~ Allele_Count, data = cloneSizes_allClones) #variance is not homogeneous
#res.aov <- aov(Clone_Size ~ Allele_Count, data = cloneSizes_allClones) #One-Way ANOVA
#summary(res.aov)
#plot(res.aov, 1) #Data contains outliers
#plot(res.aov, 2) #Data are non-normal distribution

## Since ANOVA assumptions aren't met, use Welch's ANOVA with Games-Howell post-test for multiple comparisons
#res.waov <- oneway.test(Clone_Size ~ Allele_Count, data = summary_CloneSize, var.equal = FALSE) #p-value = 2.095e-13
jmv.aov <- anovaOneW(data = cloneSizes_allClones, deps = Clone_Size, group = Allele_Count, welchs = TRUE, norm = FALSE, eqv = FALSE, phMethod = "gamesHowell")
jmv.aov

# Density Plot
myColors <- c("paleturquoise4", "lightslateblue", "gray87")

ggplot(cloneSizes_allClones, (aes(x=Clone_Size, fill=Allele_Count))) +
  geom_density(color = "grey", alpha = 1, adjust=1.5) +
  scale_fill_manual(values = myColors) +
  theme_bw() +
  scale_x_log10() +
  facet_wrap(~Allele_Count, labeller = labeller(Allele_Count = c("dual.TCRa" = "Dual TCRa (p = 0.0023)", 
                                                                 "dual.TCRb" = "Dual TCRb (p < 0.0001)", 
                                                                 "single.TCR" = "Single TCR"))) +
  xlab("Clone Size") +
  ggtitle("Clone size distribution (all samples combined)") +
  theme(legend.position="none", 
        panel.spacing = unit(0.1, "lines"), 
        strip.text.x = element_text(size = 8), 
        plot.title = element_text(hjust = 0.5, size = 14))
```  
    
  (p-value indicates difference in mean clone size compared to single TCR population as computed by Welch's One-Way ANOVA with Games-Howell posttest)  


#### By Tissue Type
Interestingly, the mean clone size of dual expressing T cell clones in tumor is greater than the mean clone size of single TCR expressing clones (1.7X and 3.7X TCRa and TCRb, respectively). In other words, dual expressing clones in the tumor are more expanded compared to single expressing clones. Further, The mean clone size is larger in tissues (normal or cancer) than in peripheral blood. In future sections, analysis will combine normal and cancer tissues when comparing with peripheral blood.

```{r blood, echo=FALSE}
# Peripheral blood
uniq_p <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "P")
uniq_pa <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "P" & unique_tcr_annot$allele_count == "Dual TCRa")
uniq_pb <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "P" & unique_tcr_annot$allele_count == "Dual TCRb")
uniq_ps <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "P" & unique_tcr_annot$allele_count == "Single TCR")
summary_p <- lapply(list("dual TCRa" = uniq_pa$cloneSize, 
                       "dual TCRb" = uniq_pb$cloneSize, 
                       "single TCR" = uniq_ps$cloneSize),
                  summary)
summary_p

# Peripheral blood ANOVA between allele count
densityData <- prepDensity(dfA = uniq_pa, dfB = uniq_pb, dfS = uniq_ps)

# ANOVA
jmv_p <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv_p
```

```{r normal, echo=FALSE}
# Normal tissue
uniq_n <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "N")
uniq_na <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "N" & unique_tcr_annot$allele_count == "Dual TCRa")
uniq_nb <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "N" & unique_tcr_annot$allele_count == "Dual TCRb")
uniq_ns <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "N" & unique_tcr_annot$allele_count == "Single TCR")
summary_n <- lapply(list("dual TCRa" = uniq_na$cloneSize, 
                       "dual TCRb" = uniq_nb$cloneSize, 
                       "single TCR" = uniq_ns$cloneSize),
                  summary)
summary_n

# Normal tissue ANOVA between allele count
densityData <- prepDensity(dfA = uniq_na, dfB = uniq_nb, dfS = uniq_ns)

# ANOVA
jmv.aov <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv.aov
```

```{r tumor, echo=FALSE}
# Tumor tissue
uniq_t <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "T")
uniq_ta <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "T" & unique_tcr_annot$allele_count == "Dual TCRa")
uniq_tb <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "T" & unique_tcr_annot$allele_count == "Dual TCRb")
uniq_ts <- subset(unique_tcr_annot, subset = unique_tcr_annot$loc == "T" & unique_tcr_annot$allele_count == "Single TCR")
summary_t <- lapply(list("dual TCRa" = uniq_ta$cloneSize, 
                       "dual TCRb" = uniq_tb$cloneSize, 
                       "single TCR" = uniq_ts$cloneSize),
                  summary)
summary_t

# Tumor tissue ANOVA between allele count
densityData <- prepDensity(dfA = uniq_ta, dfB = uniq_tb, dfS = uniq_ts)

# ANOVA
jmv.aov <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv.aov
```

##### Statistical Comparisons Between Tissue Types
-The mean clone size of dual TCRa expressing clones is greater in tissue (normal and tumor) compared to the peripheral blood, however, mean clone size is not significantly different between normal and tumor tissues.  

-The mean clone size of dual TCRb expressing clones is greater in tissue (normal and tumor) compared to the peripheral blood, however, mean clone size is not significantly different between normal and tumor tissues.  

-The mean clone size of single TCR expressing clones is greater in tissue (normal and tumor) compared to the peripheral blood, however, mean clone size is not significantly different between normal and tumor tissues. It is interesting to note that the mean clone size in tumor versus normal tissue is approaching significance (p=0.08).  

```{r Fig.2A , echo=FALSE, warning=FALSE, fig.show='hold', out.width="50%"}
# Graph mean clone size by tissue type
unique_tcr_annot_noDoublet %>%
  mutate(allele_count=factor(allele_count, levels=c("Dual TCRa", "Dual TCRb", "Single TCR"))) %>%
  mutate(loc = fct_reorder(loc, cloneSize)) %>%
  mutate(loc = factor(loc, levels=c("P", "N", "T"))) %>%
  ggplot(aes(fill=allele_count, y=cloneSize, x=loc)) +
  geom_bar(stat="summary", fun = "mean", alpha=0.9, width=0.8) +
  facet_wrap(~allele_count) +
  scale_fill_manual(values = myColors) +
  theme_minimal() +
  xlab("Tissue") +
  ylab("Mean Clone Size") +
  ggtitle("Mean Clone Size by Tissue Type") +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text.x = element_blank())
ggsave("Fig.2A Mean Clone Size by Tissue Type.pdf", device = "pdf", path = "output", width = 5, height = 3)
```

```{r between tissues ANOVA}
# Dual TCRa
uniq_a <- bind_rows(uniq_pa, uniq_na, uniq_ta)
jmv_a <- anovaOneW(data = uniq_a, deps = cloneSize, group = loc, phMethod = "gamesHowell")
jmv_a  # pvalue T vs N (0.9296394)

# Dual TCRb
uniq_b <- bind_rows(uniq_pb, uniq_nb, uniq_tb)
jmv_b <- anovaOneW(data = uniq_b, deps = cloneSize, group = loc, phMethod = "gamesHowell")
jmv_b # pvalue T vs N (0.9349664)

# Single TCR
uniq_s <- bind_rows(uniq_ps, uniq_ns, uniq_ts)
jmv_s <- anovaOneW(data = uniq_s, deps = cloneSize, group = loc, phMethod = "gamesHowell")
jmv_s # pvalue T vs N (0.0786585)
```

##### Tables
```{r summary stats overall, echo = FALSE, fig.show='hold', out.width="50%"}
summaryCloneSize <- function(df, locAsString) {
  # This function calculates mean clone size and n for each TCR allele group split by location
  
  a <- subset(df, subset = df$allele_count == "Dual TCRa" & df$loc == locAsString)
  b <- subset(df, subset = df$allele_count == "Dual TCRb" & df$loc == locAsString)
  s <- subset(df, subset = df$allele_count == "Single TCR" & df$loc == locAsString)
  this.list <- list(a,b,s)
  
  this.df <- data.frame(matrix(nrow = 3, ncol = 3))
  colnames(this.df) <- c("allele_count", "mean_cloneSize", "N_uniqueClones")
  
  for (i in 1:length(this.list)) {
    mean <- format(round(mean(this.list[[i]]$cloneSize), 2), nsmall = 1)
    n <- dim(this.list[[i]])[1]
    this.df[i, "allele_count"] <- names(table(this.list[[i]]$allele_count)[1])
    this.df[i, "mean_cloneSize"] <- mean
    this.df[i, "N_uniqueClones"] <- n
  }
  return(this.df)
}

# Calculating mean cloneSize and counting N by allele count by tissue location of all clones
table_t <- summaryCloneSize(unique_tcr_annot, "T")
table_n <- summaryCloneSize(unique_tcr_annot, "N")
table_p <- summaryCloneSize(unique_tcr_annot, "P")

# Calculating mean cloneSize and counting N by allele count by tissue location of clonal clones
clonal_only_tcr <- subset(unique_tcr_annot, subset = unique_tcr_annot$cloneStatus=="Clonal") 
table_ct <- summaryCloneSize(clonal_only_tcr, "T")
table_cn <- summaryCloneSize(clonal_only_tcr, "N")
table_cp <- summaryCloneSize(clonal_only_tcr, "P")

n_table <- cbind(" " = table_t$allele_count,
                 "Peripheral Blood (P)" = table_p$N_uniqueClones,
                 "Normal (N)" = table_n$N_uniqueClones,
                 "Tumor (T)" = table_t$N_uniqueClones)

mean_table <- cbind(" " = table_t$allele_count,
                    "Peripheral Blood (P)" = table_p$mean_cloneSize,
                    "Normal (N)" = table_n$mean_cloneSize,
                    "Tumor (T)" = table_t$mean_cloneSize)

cn_table <- cbind(" " = table_ct$allele_count,
                  "Peripheral Blood (P)" = table_cp$N_uniqueClones,
                  "Normal (N)" = table_cn$N_uniqueClones,
                  "Tumor (T)" = table_ct$N_uniqueClones)

cmean_table <- cbind(" " = table_ct$allele_count,
                     "Peripheral Blood (P)" = table_cp$mean_cloneSize,
                     "Normal (N)" = table_cn$mean_cloneSize,
                     "Tumor (T)" = table_ct$mean_cloneSize)

# Generate Tables
knitr::kable(n_table, caption = "Table 1. Number of unique clones")
knitr::kable(mean_table, caption = "Table 2. Mean clone size of unique clones")

knitr::kable(cn_table, caption = "Table 3. Number of unique, clonal clones")
knitr::kable(cmean_table, caption = "Table 4. Mean clone size of unique, clonal clones")
```


#### By T cell subset
Because dual TCR cell frequencies and mean expansion differed greatly between peripheral blood and tissues, tissue (combined normal and tumor) and peripheral blood are evaluated separately here.  
  
**Summary of results in this section:**  
* In all cases, CD4+ clones were less clonally expanded than CD8+ clones.  
* In tissues, the mean clone size of unique CD4+ T cell clones was not significantly increased compared to the single TCR subpopulation (p = 0.08).  
* In tissues, mean clone size of unique CD8+ T cell clones was significantly increased compared to the single TCR subpopulation (p = 0.0058).  
* the number of expanded unique dual TCRa (1.9X) and dual TCRb (4.1X) expressing CD8+ clones was significantly greater compared to single TCR expressing CD8+ clones. 

```{r blood by subtype, echo=FALSE, fig.show='hold', out.width="50%"}
uniq_blood <- subset(unique_tcr_annot_noDoublet, subset = unique_tcr_annot_noDoublet$loc == "P")

# Graph mean blood clone size by CD4/CD8
uniq_blood %>%
  mutate(allele_count=factor(allele_count, levels=c("Dual TCRa", "Dual TCRb", "Single TCR"))) %>%
  mutate(stype = fct_reorder(stype, cloneSize)) %>%
  mutate(stype = factor(stype, levels=c("CD4", "CD8"))) %>%
  ggplot(aes(fill=allele_count, y=cloneSize, x=stype)) +
  geom_bar(stat="summary", fun = "mean", alpha=0.9, width=0.8) +
  facet_wrap(~allele_count) +
  scale_fill_manual(values = myColors) +
  theme_minimal() +
  xlab("T cell subset") +
  ylab("Mean Clone Size") +
  ggtitle("Mean clone size by T cell subset (peripheral blood)") +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text.x = element_blank())
ggsave("Fig.2B Mean Clone Size by T cell subset (blood).pdf", device = "pdf", path = "output", width = 5, height = 3)
```

```{r ANOVAs blood by subtype, echo=FALSE, fig.show='hold', out.width="50%"}
# Blood CD4+ ANOVA between allele count
uniq_cd4a <- subset(uniq_blood, subset = uniq_blood$stype == "CD4" & uniq_blood$allele_count == "Dual TCRa")
uniq_cd4b <- subset(uniq_blood, subset = uniq_blood$stype == "CD4" & uniq_blood$allele_count == "Dual TCRb")
uniq_cd4s <- subset(uniq_blood, subset = uniq_blood$stype == "CD4" & uniq_blood$allele_count == "Single TCR")

stats_cd4 <- lapply(list("dual TCRa" = uniq_cd4a$cloneSize, 
                         "dual TCRb" = uniq_cd4b$cloneSize, 
                         "single TCR" = uniq_cd4s$cloneSize),
                    summary)
stats_cd4

densityData <- prepDensity(dfA = uniq_cd4a, dfB = uniq_cd4b, dfS = uniq_cd4s)

# CD4+ ANOVA
jmv.aov <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv.aov


# Blood CD8+ ANOVA between allele count
uniq_cd8a <- subset(uniq_blood, subset = uniq_blood$stype == "CD8" & uniq_blood$allele_count == "Dual TCRa")
uniq_cd8b <- subset(uniq_blood, subset = uniq_blood$stype == "CD8" & uniq_blood$allele_count == "Dual TCRb")
uniq_cd8s <- subset(uniq_blood, subset = uniq_blood$stype == "CD8" & uniq_blood$allele_count == "Single TCR")

stats_cd8 <- lapply(list("dual TCRa" = uniq_cd8a$cloneSize, 
                         "dual TCRb" = uniq_cd8b$cloneSize, 
                         "single TCR" = uniq_cd8s$cloneSize),
                    summary)
stats_cd8

densityData <- prepDensity(dfA = uniq_cd8a, dfB = uniq_cd8b, dfS = uniq_cd8s)

# CD8+ ANOVA
jmv.aov <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv.aov

```

```{r tissue by subtype, echo=FALSE, fig.show='hold', out.width="50%"}
# Tissue subset unique CD4 clones by allele subtype
uniq_tissue <- subset(unique_tcr_annot_noDoublet, subset = unique_tcr_annot_noDoublet$loc != "P")

# Graph mean tissue clone size by CD4/CD8
uniq_tissue %>%
  mutate(allele_count=factor(allele_count, levels=c("Dual TCRa", "Dual TCRb", "Single TCR"))) %>%
  mutate(stype = fct_reorder(stype, cloneSize)) %>%
  mutate(stype = factor(stype, levels=c("CD4", "CD8"))) %>%
  ggplot(aes(fill=allele_count, y=cloneSize, x=stype)) +
  geom_bar(stat="summary", fun = "mean", alpha=0.9, width=0.8) +
  facet_wrap(~allele_count) +
  scale_fill_manual(values = myColors) +
  theme_minimal() +
  xlab("T cell subset") +
  ylab("Mean Clone Size") +
  ggtitle("Mean clone size in tissue by T cell subset") +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text.x = element_blank())
ggsave("Fig.2C Mean Clone Size by T cell subset (tissue).pdf", device = "pdf", path = "output", width = 5, height = 3)
```

```{r ANOVAs tissue by subtype, echo=FALSE, fig.show='hold', out.width="50%"}
# Tissue CD4+ ANOVA between allele count
uniq_cd4a <- subset(uniq_tissue, subset = uniq_tissue$stype == "CD4" & uniq_tissue$allele_count == "Dual TCRa")
uniq_cd4b <- subset(uniq_tissue, subset = uniq_tissue$stype == "CD4" & uniq_tissue$allele_count == "Dual TCRb")
uniq_cd4s <- subset(uniq_tissue, subset = uniq_tissue$stype == "CD4" & uniq_tissue$allele_count == "Single TCR")

stats_cd4 <- lapply(list("dual TCRa" = uniq_cd4a$cloneSize, 
                         "dual TCRb" = uniq_cd4b$cloneSize, 
                         "single TCR" = uniq_cd4s$cloneSize),
                    summary)
stats_cd4

densityData <- prepDensity(dfA = uniq_cd4a, dfB = uniq_cd4b, dfS = uniq_cd4s)

# CD4+ ANOVA
jmv.aov <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv.aov

# Tissue CD8+ ANOVA between allele count
uniq_cd8a <- subset(uniq_tissue, subset = uniq_tissue$stype == "CD8" & uniq_tissue$allele_count == "Dual TCRa")
uniq_cd8b <- subset(uniq_tissue, subset = uniq_tissue$stype == "CD8" & uniq_tissue$allele_count == "Dual TCRb")
uniq_cd8s <- subset(uniq_tissue, subset = uniq_tissue$stype == "CD8" & uniq_tissue$allele_count == "Single TCR")

stats_cd8 <- lapply(list("dual TCRa" = uniq_cd8a$cloneSize, 
                         "dual TCRb" = uniq_cd8b$cloneSize, 
                         "single TCR" = uniq_cd8s$cloneSize),
                    summary)
stats_cd8

densityData <- prepDensity(dfA = uniq_cd8a, dfB = uniq_cd8b, dfS = uniq_cd8s)

# CD8+ ANOVA
jmv.aov <- anovaOneW(data = densityData, deps = Clone_Size, group = Allele_Count, phMethod = "gamesHowell")
jmv.aov
```




#### By T cell cluster phenotype

``` {r calc_aovs_byLoc function}
calc_aovs_byLoc <- function (df, myList, allele) {
  # This function performs Welch's ANOVA on cloneSize by location by cluster ID
  for (i in 1:length(myList)) {
    this.cluster <- clusterNames[i]
    this.subset <- subset(df, subset = cluster.name == this.cluster & allele_count == allele)
    
    tryCatch(
      expr = {
        this.aov <- anovaOneW(data = this.subset, deps = cloneSize, group = loc, phMethod = "gamesHowell")
      },
      error = function(e) {
        message(paste0("Unsuccesful attempt to evaluate cluster ", this.cluster))
        print(e)
      },
      finally = {
        print(this.cluster)
        print(this.aov)
      })
  }
}
```

Many T cell clusters were similar in clonal expansion between tumor and normal tissues. A few, unsurprisingly, were different (more expanded Tregs in tumor, for example). 

Clusters with significantly different **dual TCRa** mean clone size between tumor and normal tissue include:  
CD4.c17(IFNG+ Tfh/Th1), pvalue = 0.0003910  
CD4.c20(TNFRSF9+ Treg), pvalue = 0.0007275  
CD8.c12(terminal Tex), pvalue = 0.0271305  

Clusters with significantly different **dual TCRb** mean clone size between tumor and normal tissue include:  
CD4.c20(TNFRSF9+ Treg), pvalue = 0.0002099  

Clusters with significantly different **single TCR** mean clone size between tumor and normal tissue include:  
CD4.c05(TNF+ T), pvalue = 0.0065341  
CD4.c14(CCR6+ Th17), pvalue = 0.0388218  
CD4.c16(IL21+ Tfh), pvalue = 0.0004987  
CD4.c19(S1PR1+ Treg), pvalue = 0.0429657  
CD4.c20(TNFRSF9+ Treg), pvalue = 0.0000009  
CD4.c21(ISG+ Treg), pvalue = 0.0002077  
CD4.c22(ISG+ Th), pvalue = 0.0148291  
CD4.c23(NME1+CCR4- T), pvalue = 0.0069151  
CD8.c03(uncharacterized), pvalue = 0.0267355  

``` {r ANOVA by clusterID, warning=FALSE, error=FALSE, message=FALSE}
# Get a list of cluster IDs
clusterNames <- data.frame(table(unique_tcr_annot$cluster.name))$Var1

#Dual TCRa
calc_aovs_byLoc(unique_tcr_annot, clusterNames, allele = "Dual TCRa")
#Dual TCRb
calc_aovs_byLoc(unique_tcr_annot, clusterNames, allele = "Dual TCRb")
#Single TCR
calc_aovs_byLoc(unique_tcr_annot, clusterNames, allele = "Single TCR")
```

#### By allele count in tissue 
Initial analysis indicated that dual TCR clones were more expanded (larger mean clone size) in tissue compared to single TCR clones, regardless of whether normal or tumor. To determine whether certain T cell clusters were more or less expanded, the mean clone size was calculated for dual and single TCR clones for each cluster.  

Clusters with significantly different mean clone size > 2-fold include:  

Dual TCRb vs single:  
CD8.c05(GZMK+ early Tem)  
CD8.c10(ZNF683+CXCR6+ Trm)  
CD8.c12(terminal Tex)  
CD8.c17(NME1+ T)  

Thus, CD8+ T cells with highly differentiated phenotypes (memory and exhausted) that express more than one TCR beta chain have a higher mean clone size than CD8+ T cells expressing only one TCR beta chain. This could result from a propensity of dual expressing cells to differentiate into tissue resident memory cells, greater clonal expansion within tissue, or heightened ability to migrate into tissues and be retained there.

``` {r}
# Perform Welch's ANOVA on cloneSize by allele count by cluster ID
calc_aovs_byAllele <- function (df, myList) {
  for (i in 1:length(myList)) {
    this.cluster <- clusterNames[i]
    this.subset <- subset(df, 
                          subset = cluster.name == this.cluster & allele_count != "Dual TCRa and TCRb" & loc != "P")
    this.aov <- anovaOneW(data = this.subset, deps = cloneSize, 
                          group = allele_count, phMethod = "gamesHowell")
    print(this.cluster)
    print(this.aov)
  }
}
```

```{r ANOVA by allele}
# Get a list of cluster IDs
clusterNames <- data.frame(table(unique_tcr_annot$cluster.name))$Var1

calc_aovs_byAllele(unique_tcr_annot, clusterNames)
```

##### Tables of significant findings (by allele count)
```{r, echo=FALSE}
# Create table data of significant findings
calcRatio <- function (df, myList, allele) {
  newdf <- data.frame(matrix(ncol = 5, nrow = length(myList)))
  colnames(newdf) <- c("clusterID", paste0("mean_dual", allele), "mean_single", "ratio", "pvalue") 
  
  for (i in 1:length(myList)) {
    this.cluster <- names(myList)[i]
    this.subS <- subset(df, subset = cluster.name == this.cluster & allele_count == "Single TCR" & loc != "P")
    
    if (allele == "A") {
      this.sub <- subset(df, subset = cluster.name == this.cluster & allele_count == "Dual TCRa" & loc != "P")
    } else if (allele == "B") {
      this.sub <- subset(df, subset = cluster.name == this.cluster & allele_count == "Dual TCRb" & loc != "P")
    }

    newdf[i,"clusterID"] <- this.cluster
    newdf[i,"pvalue"] <- myList[i]
    newdf[i, "mean_single"] <- mean(this.subS$cloneSize)
    newdf[i, paste0("mean_dual", allele)] <- mean(this.sub$cloneSize)
    newdf[i, "ratio"] <- mean(this.sub$cloneSize) / mean(this.subS$cloneSize)
  }
  return(newdf)
}


sigClusters_a <- list("CD4.c07(TIMP1+ Tm)" = 0.0000144,
                      "CD4.c12(GZMK+ Tem)" = 0.0387535,
                      "CD4.c21(ISG+ Treg)" = 0.0003878)

sigClusters_b <- list("CD4.c01(Tn)" = 0.0052142,
                      "CD4.c04(IL7R- Tn)" = 0.0001980,
                      "CD4.c05(TNF+ T)" = 0.0000001,
                      "CD4.c07(TIMP1+ Tm)" = 0.0036434,
                      "CD4.c20(TNFRSF9+ Treg)" = 0.0006711,
                      "CD8.c05(GZMK+ early Tem)" = 0.0240358,
                      "CD8.c10(ZNF683+CXCR6+ Trm)" = 0.0001229,
                      "CD8.c12(terminal Tex)" = 0.0063197,
                      "CD8.c17(NME1+ T)" = 0.0022428)

ratios_a <- calcRatio(unique_tcr_annot, sigClusters_a, "A")
ratios_b <- calcRatio(unique_tcr_annot, sigClusters_b, "B")

# Generate Tables
knitr::kable(ratios_a, caption = "Table 7. Clusters where mean clone size of dual TCRa expressors differs")
knitr::kable(ratios_b, caption = "Table 8. Clusters where mean clone size of dual TCRb expressors differs")
```

```{r Fig.3}
# Visualize most convincing comparisons

pairwiseReshape <- function(df) {
  newDF <- data.frame(matrix(nrow = 2*dim(df)[1], ncol = 3))
  colnames(newDF) <- c("clusterID", "group", "value")
  
  newDF$clusterID <- rep(df$clusterID, 2)
  newDF$group <- c(rep(colnames(df)[2], dim(df)[1]),
                   rep(colnames(df)[3], dim(df)[1]))
  newDF$value <- c(df[,2], df[,3])
  return(newDF)
}

convincing <- pairwiseReshape(ratios_b[6:8,])

ggplot(convincing, aes(x = group, y=value, fill = group)) +
  geom_bar(stat="identity", alpha = 0.9, width = 0.8) +
  facet_wrap(~clusterID) +
  scale_fill_manual(values = c("lightslateblue", "gray87")) +
  theme_minimal() +
  xlab("") +
  ylab("Mean Clone Size") +
  ggtitle("Mean clone size within indicated cluster") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.text.x = element_blank())
ggsave("Fig.3 Mean clone size within indicated cluster.pdf", device = "pdf", path = "output", width = 6, height = 4)
```



### TR[A|B]V gene usage

```{r functions}

firstWord <- function(my.string) {
    unlist(strsplit(my.string, "_"))[1]
}

secondWord <-function(my.string) {
  my.string <- toString(my.string)
  as.numeric(str_extract(my.string, "\\d+"))
}

annotTRV <- function(df) {
  va1 <- sapply(sub(pattern = "\\|", "_", df$Identifier.A1), firstWord)
  va2 <- sapply(sub(pattern = "\\|", "_", df$Identifier.A2), firstWord)
  vb1 <- sapply(sub(pattern = "\\|", "_", df$Identifier.B1), firstWord)
  vb2 <- sapply(sub(pattern = "\\|", "_", df$Identifier.B2), firstWord)
  df$va1 <- va1
  df$va2 <- va2
  df$vb1 <- vb1
  df$vb2 <- vb2
  return(df)
}
countTRV <- function(df, chain) {
  
  if(chain == "beta") {
    v1 <- data.frame(table(df$vb1))
    v2 <- data.frame(table(df$vb2))
    geneList <- unique(c(unique(uniq_tcr$vb1), unique(uniq_tcr$vb2)))
  } else if (chain == "alpha") {
    v1 <- data.frame(table(df$va1))
    v2 <- data.frame(table(df$va2))
    geneList <- unique(c(unique(uniq_tcr$va1), unique(uniq_tcr$va2)))
  }
  
  geneList <- geneList[!is.na(geneList)]
  
  colnames(v1) <- c("gene", "clone_count_v1")
  colnames(v2) <- c("gene", "clone_count_v2")
  vb <- full_join(v1, v2, by = "gene")
  vb <- vb %>% replace(is.na(.), 0)
  vb$comb_cloneCount <- vb$clone_count_v1 + vb$clone_count_v2
  
  masterDF <- data.frame("gene" = geneList)
  masterDF <- full_join(masterDF, vb, by = "gene") %>% replace(is.na(.), 0)
  masterDF$order <- sapply(masterDF$gene, secondWord)
  masterDF <- masterDF[order(masterDF$order),]
  
  newDF <- data.frame("gene" = masterDF$gene,
                      "cloneCount"= masterDF$comb_cloneCount)
  newDF <- subset(newDF, subset = newDF$gene != "-")
  countTotal <- sum(newDF$cloneCount)
  newDF$cloneProportion <- newDF$cloneCount/countTotal
  
  
  return(newDF)
}
```


```{r create TRBV dataframes, echo=FALSE, fig.dim=c(6,10), fig.show="hold", out.width="30%"}
uniq_tcr <- annotTRV(unique_tcr_annot_noDoublet)

trbv_a <- countTRV(subset(uniq_tcr, subset = uniq_tcr$allele_count == "Dual TCRa"), chain = "beta")
trbv_b <- countTRV(subset(uniq_tcr, subset = uniq_tcr$allele_count == "Dual TCRb"), chain = "beta")
trbv_s <- countTRV(subset(uniq_tcr, subset = uniq_tcr$allele_count == "Single TCR"), chain = "beta")

trav_a <- countTRV(subset(uniq_tcr, subset = uniq_tcr$allele_count == "Dual TCRa"), chain = "alpha")
trav_b <- countTRV(subset(uniq_tcr, subset = uniq_tcr$allele_count == "Dual TCRb"), chain = "alpha")
trav_s <- countTRV(subset(uniq_tcr, subset = uniq_tcr$allele_count == "Single TCR"), chain = "alpha")

barplot(height = trbv_a$cloneProportion, names=trbv_a$gene,
        col = "paleturquoise4",
        las = 2,
        cex.names = 0.5,
        xlim = c(0, 0.12),
        xlab = "Proportion",
        horiz = TRUE,
        main = "TRBV gene usage by cells expressing two alpha chains")

barplot(height = trbv_b$cloneProportion, names=trbv_b$gene,
        col = "lightslateblue",
        las = 2,
        cex.names = 0.5,
        xlim = c(0, 0.12),
        xlab = "Proportion",
        horiz = TRUE,
        main = "TRBV gene usage by cells expressing two beta chains")

barplot(height = trbv_s$cloneProportion, names=trbv_s$gene,
        col = "gray87",
        las = 2,
        cex.names = 0.5,
        xlim = c(0, 0.12),
        xlab = "Proportion",
        horiz = TRUE,
        main = "TRBV gene usage by cells expressing a single TCR")

barplot(height = trav_a$cloneProportion, names=trav_a$gene,
        col = "paleturquoise4",
        las = 2,
        cex.names = 0.5,
        xlim = c(0, 0.06),
        xlab = "Proportion",
        horiz = TRUE,
        main = "TRAV gene usage by cells expressing two alpha chains")

barplot(height = trav_b$cloneProportion, names=trav_b$gene,
        col = "lightslateblue",
        las = 2,
        cex.names = 0.5,
        xlim = c(0, 0.06),
        xlab = "Proportion",
        horiz = TRUE,
        main = "TRAV gene usage by cells expressing two beta chains")

barplot(height = trav_s$cloneProportion, names=trav_s$gene,
        col = "gray87",
        las = 2,
        cex.names = 0.5,
        xlim = c(0, 0.06),
        xlab = "Proportion",
        horiz = TRUE,
        main = "TRAV gene usage by cells expressing a single TCR")
```





## Session Information
```{r session info, results="markup", include=TRUE}
sessionInfo()
```