# Dual-TCR-in-cancer
In 2021, Zheng et. al published a [pan-cancer single cell atlas](https://www.science.org/doi/10.1126/science.abe6474?url_ver=Z39.88-2003&rfr_id=ori:rid:crossref.org&rfr_dat=cr_pub%20%200pubmed#supplementary-materials) with gene expression data from 316 cancer patients and TCRseq data from 87 cancer patients. Tissues included in studies include peripheral blood, normal tissue, and tumor tissue. Authors used this data for repertoire analysis comparisons and to identify the routes to T cell exhaustion. However, authors did not evaluate the frequency of cells expressing dual productive TCR sequences. The Zhang lab provides objects containing pre-processed [data](https://zenodo.org/record/5461803#.Y5D4sOzMKAk). These were downloaded and read into R. Gene expression data were integrated by the original authors using the [Harmony](https://www.nature.com/articles/s41592-019-0619-0) correction algorithm. Here, we re-analyze the Zheng Atlas TCRseq data to compute the frequencies of dual TCR expressing T cells in cancer and identify differences in clonal expansion between tissue type and T cell phenotype.

 
